Mathlib/Topology/Algebra/Group/OpenMapping.lean

```diff
@@ -0,0 +1,123 @@
+-/
+
+open scoped Topology Pointwise
+
+open MulAction Set Function
+
+variable {G X : Type*} [TopologicalSpace G] [TopologicalSpace X]
+  [Group G] [TopologicalGroup G] [MulAction G X]
+  [SigmaCompactSpace G] [BaireSpace X] [T2Space X]
+  [ContinuousSMul G X] [IsPretransitive G X]
+
+/-- Consider a sigma-compact group acting continuously and transitively on a Baire space. Then
+the orbit map is open around the identity. It follows in `isOpenMap_smul_of_sigmaCompact` that it
+is open around any point. -/
+@[to_additive "Consider a sigma-compact additive group acting continuously and transitively on a
+Baire space. Then the orbit map is open around zero. It follows in
+`isOpenMap_vadd_of_sigmaCompact` that it is open around any point."]
+theorem smul_singleton_mem_nhds_of_sigmaCompact
+    {U : Set G} (hU : U ∈ 𝓝 1) (x : X) : U • {x} ∈ 𝓝 x := by
+  /- Consider a small closed neighborhood `V` of the identity. Then the space is covered by
+  countably many translates of `V`, say `gᵢ V`. Let also `Kₙ` be a sequence of compact sets covering
+  the space. Then the image of `Kₙ ∩ gᵢ V` in the orbit is compact, and their unions covers the
+  space. By Baire, one of them has nonempty interior. Then `gᵢ V • x` has nonempty interior, and
+  so does `V • x`. Its interior contains a point `g' x` with `g' ∈ V`. Then `g'⁻¹ • V • x` contains
+  a neighborhood of `x`, and it is included in `V⁻¹ • V • x`, which is itself contained in `U • x`
+  if `V` is small enough. -/
+  obtain ⟨V, V_mem, V_closed, V_symm, VU⟩ : ∃ V ∈ 𝓝 (1 : G), IsClosed V ∧ V⁻¹ = V ∧ V * V ⊆ U :=
+    exists_nhds_one_isClosed_inv_eq_mul_subset hU
+  obtain ⟨s, s_count, hs⟩ : ∃ (s : Set G), s.Countable ∧ ⋃ g ∈ s, g • V = univ := by
+    apply countable_cover_nhds_of_sigma_compact (fun g ↦ ?_)
+    convert smul_mem_nhds g V_mem
+    simp only [smul_eq_mul, mul_one]
+  let K : ℕ → Set G := compactCovering G
+  let F : ℕ × s → Set X := fun p ↦ (K p.1 ∩ (p.2 : G) • V) • ({x} : Set X)
+  obtain ⟨⟨n, ⟨g, hg⟩⟩, hi⟩ : ∃ i, (interior (F i)).Nonempty := by
+    have : Nonempty X := ⟨x⟩
+    have : Encodable s := Countable.toEncodable s_count
+    apply nonempty_interior_of_iUnion_of_closed
+    · rintro ⟨n, ⟨g, hg⟩⟩
+      apply IsCompact.isClosed
+      suffices H : IsCompact ((fun (g : G) ↦ g • x) '' (K n ∩ g • V)) by
+        simpa only [smul_singleton] using H
+      apply IsCompact.image
+      · exact (isCompact_compactCovering G n).inter_right (V_closed.smul g)
+      · exact continuous_id.smul continuous_const
+    · apply eq_univ_iff_forall.2 (fun y ↦ ?_)
+      obtain ⟨h, rfl⟩ : ∃ h, h • x = y := exists_smul_eq G x y
+      obtain ⟨n, hn⟩ : ∃ n, h ∈ K n := exists_mem_compactCovering h
+      obtain ⟨g, gs, hg⟩ : ∃ g ∈ s, h ∈ g • V := exists_set_mem_of_union_eq_top s _ hs _
+      simp only [smul_singleton, mem_iUnion, mem_image, mem_inter_iff, Prod.exists, Subtype.exists,
+        exists_prop]
+      exact ⟨n, g, gs, h, ⟨hn, hg⟩, rfl⟩
```

Comment:
I haven't tried but I think this could be simpler if you just stick to manipulating unions. I think combining [Set.iUnion_inter](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Set/Lattice.html#Set.iUnion_inter), [iSup_prod](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Order/CompleteLattice.html#iSup_prod) could get you there by rewriting only.

Reply:
Oh actually we have no formulation of transitivity well-suited for that, so you'd have to add that too...

Reply:
I feel like a direct proof where you just follow your nose is probably the easiest to read. So even if there was a good sequence of rewrites, I think I'd rather stick to the current proof.